synapse/util/caches/expiringcache.py

```diff
@@ -210,7 +212,7 @@ def set_cache_factor(self, factor: float) -> bool:
         return False
 
 
-@attr.s(slots=True)
+@attr.s(slots=True, auto_attribs=True)
 class _CacheEntry:
-    time = attr.ib(type=int)
-    value = attr.ib()
+    time: int
+    value: Any
```

Comment:
Right, I think this can't be a generic yet so I made it `Any`, which I think is what we've done for other situations. It is rather meh though. 👎

I forget if we just need newer versions of attrs or Python or what is blocking us here though. We should probably file an issue somewhere about this...

Reply:
https://github.com/python-attrs/attrs/issues/313 tracks the issue. I wonder if it's something we could take a look at---we've certainly been bitten by it frequently.

That said, on the occasions we've been bitten I don't think we're using any of the `attrs` machinery at all. I reckon a plain [`dataclass`](https://docs.python.org/3/library/dataclasses.html?highlight=dataclass#dataclasses.dataclass) with `slots=True` would give us exactly what we want for this case. I suspect the team may have Opinions though.

I think this change is fine for this PR!